Mathlib/Algebra/Category/ModuleCat/Presheaf/Pullback.lean

```diff
@@ -0,0 +1,95 @@
+/-
+Copyright (c) 2024 Joël Riou. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joël Riou
+-/
+import Mathlib.Algebra.Category.ModuleCat.Presheaf.Generator
+import Mathlib.Algebra.Category.ModuleCat.Presheaf.Pushforward
+import Mathlib.CategoryTheory.Adjunction.PartialAdjoint
+
+/-!
+# Pullback of presheaves of modules
+
+Let `F : C ⥤ D` be a functor, `R : Dᵒᵖ ⥤ RingCat` and `S : Cᵒᵖ ⥤ RingCat` be presheaves
+of rings, and `φ : S ⟶ F.op ⋙ R` be a morphism of presheaves of rings,
+we introduce the pullback functor `pullback : PresheafOfModules S ⥤ PresheafOfModules R`
+as the left adjoint of `pushforward : PresheafOfModules R ⥤ PresheafOfModules S`.
+The existence of this left adjoint functor is obtained under suitable universe assumptions.
+
+-/
+
+universe v v₁ v₂ u₁ u₂ u
+
+open CategoryTheory Limits Opposite
+
+namespace PresheafOfModules
+
+section
+
+variable {C : Type u₁} [Category.{v₁} C] {D : Type u₂} [Category.{v₂} D]
+  {F : C ⥤ D} {R : Dᵒᵖ ⥤ RingCat.{u}} {S : Cᵒᵖ ⥤ RingCat.{u}} (φ : S ⟶ F.op ⋙ R)
+  [(pushforward.{v} φ).IsRightAdjoint]
+
+/-- The pullback functor `PresheafOfModules S ⥤ PresheafOfModules R` induced by
+a morphism of presheaves of rings `S ⟶ F.op ⋙ R`, defined as the left adjoint
+functor to the pushforward, when it exists. -/
+noncomputable def pullback : PresheafOfModules.{v} S ⥤ PresheafOfModules.{v} R :=
+  (pushforward.{v} φ).leftAdjoint
+
+/-- Given a morphism of presheaves of rings `S ⟶ F.op ⋙ R`, this is the adjunction
+between associated pullback and pushforward functors on the categories
+of presheaves of modules. -/
+noncomputable def pullbackPushforwardAdjunction : pullback.{v} φ ⊣ pushforward.{v} φ :=
+  Adjunction.ofIsRightAdjoint (pushforward φ)
+
+/-- Given a morphism of presheaves of rings `φ : S ⟶ F.op ⋙ R`, this is the property
+that the (partial) left adjoint functor of `pushforward φ` is defined
+on a certain object `M : PresheafOfModules S`. -/
+abbrev PullbackObjIsDefined : PresheafOfModules.{v} S → Prop :=
+  (pushforward φ).LeftAdjointObjIsDefined
+
+end
+
+section
+
+variable {C D : Type u} [SmallCategory C] [SmallCategory D]
+  {F : C ⥤ D} {R : Dᵒᵖ ⥤ RingCat.{u}} {S : Cᵒᵖ ⥤ RingCat.{u}} (φ : S ⟶ F.op ⋙ R)
```

Comment:
Maybe it is useful to generalize the final result to `[Category {C,D}] [EssentiallySmall {C,D}]`?

Reply:
Not for this PR. (I think it would require showing that if `F` is an equivalence and `φ` is an isomorphism, then the associated pushforward functor is an equivalence of categories.)